File: 5_longest_palindromic_substring/longestPalindrome.ts
function longestPalindrome(s: string): string {
  const n = s.length;

  let maxLen = 0;
  let res = "";

  for (let i = 0; i < n; i++) {
    let l = i;
    let r = i;

    while (l >= 0 && r < n && s[l] === s[r]) {
      let diff = r - l + 1;
      if (diff > maxLen) {
        maxLen = diff;
        res = s.slice(l, r + 1);
      }
      l -= 1;
      r += 1;
    }
    l = i;
    r = i + 1;

    while (l >= 0 && r < n && s[l] === s[r]) {
      let diff = r - l + 1;
      if (diff > maxLen) {
        maxLen = diff;
        res = s.slice(l, r + 1);
      }
      l -= 1;
      r += 1;
    }
  }

  return res;
}

File: 5_longest_palindromic_substring/longestPalindrome.c
char* longestPalindrome(char* s) {
    int maxLen = 0;
    int n = strlen(s);

    char* res;

    for (int i = 0; i < n; i++) {
        int l = i;
        int r = i;

        while (l >= 0 && r < n && s[l] == s[r]) {
            int diff = r - l + 1;
            if (diff > maxLen) {
                maxLen = diff;
                res = malloc(maxLen + 1);
                memcpy(res, &s[l], maxLen);
                res[maxLen] = '\0';
            }
            l--;
            r++;
        }

        l = i;
        r = i + 1;

        while (l >= 0 && r < n && s[l] == s[r]) {
            int diff = r - l + 1;
            if (diff > maxLen) {
                maxLen = diff;
                res = malloc(maxLen + 1);
                memcpy(res, &s[l], maxLen);
                res[maxLen] = '\0';
            }
            l--;
            r++;
        }
    }

    return res;
}